torch/csrc/jit/script/compilation_unit.h

```diff
@@ -30,7 +30,11 @@ struct SugaredValue;
 struct Resolver;
 
 using ResolverPtr = std::shared_ptr<Resolver>;
-using Self = std::function<std::shared_ptr<SugaredValue>(Value*)>;
+struct Self {
+  virtual ~Self() {}
+  virtual std::shared_ptr<SugaredValue> makeSugared(Value* v) const = 0;
+  virtual ClassTypePtr getClassType() const = 0;
```

Comment:
This is unused in the PR, which makes it hard for me to evaluate this PR on its own. I don't like having to do this because it re-introduces an annoying complexity for this self argument. I don't understand why we can't extract the type for the value (v->type()) or from the schema.